lib/ClangImporter/ClangImporter.cpp

```diff
@@ -768,73 +768,46 @@ bool ClangImporter::canReadPCH(StringRef PCHFilename) {
     return false;
 
   // FIXME: The following attempts to do an initial ReadAST invocation to verify
-  // the PCH, without affecting the existing CompilerInstance.
+  // the PCH, without causing trouble for the existing CompilerInstance.
   // Look into combining creating the ASTReader along with verification + update
   // if necessary, so that we can create and use one ASTReader in the common case
   // when there is no need for update.
-
-  CompilerInstance &CI = *Impl.Instance;
-  auto clangDiags = CompilerInstance::createDiagnostics(
-                                              new clang::DiagnosticOptions());
+  clang::CompilerInstance CI(Impl.Instance->getPCHContainerOperations(),
+                             &Impl.Instance->getModuleCache());
+  auto invocation =
+      std::make_shared<clang::CompilerInvocation>(*Impl.Invocation);
```

Comment:
Is this constructing a shared_ptr *copy* of Impl.Invocation?

Reply:
`make_shared<T>(args...)` is a wrapper for `shared_ptr<T>(new T(args))`, so this is constructing a (mostly) independent CompilerInvocation (see https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared).  The rest of the file is setting a bad example, by effectively doing `shared_ptr<T>(new T(T(args)))`.

I say "mostly" above because PreprocessorOptions::RemappedFileBuffers has raw, sometimes-owned pointers to MemoryBuffer :/.

Reply:
I think I'll clean up the other cases in a separate commit ahead of time so this is more clear.

Reply:
Actually, there was only one of those.  Cleaned up in 0a1d270d507308d51022609f183fbaad18fb3fa1.